src/alerts_utils.cc

```diff
@@ -277,87 +278,155 @@ alert_load_state(zlistx_t *alerts, const char *path, const char *filename) {
+        const char *encoded_alert_desc = fty_proto_description (alert);
+        size_t encoded_alert_desc_size = strlen (encoded_alert_desc);
+        uint8_t *alert_desc = (uint8_t *) zmalloc (1 + (5 * encoded_alert_desc_size) / 4);
+        zmq_z85_decode (alert_desc, encoded_alert_desc);
+
+        std::string alert_desc_str ((char *) alert_desc);
+        size_t alert_desc_end = alert_desc_str.find_last_not_of (" ");
+
+        fty_proto_set_description (alert, "%s", alert_desc_str.substr (0, alert_desc_end + 1).c_str ());
+        fty_proto_print (alert);
+
+        if (s_alerts_input_checks (alerts, alert)) {
+            log_warning (
+                    "Alert id (%s, %s) already read.",
+                    fty_proto_rule(alert),
+                    fty_proto_name(alert));
+        }
+        else {
+            zlistx_add_end (alerts, alert);
         }
-#else
-        frame = zmsg_encode(zmessage);
-        size = zframe_size(frame);
-#endif
-        zmsg_destroy(&zmessage);
-        assert(frame);
-        assert(size > 0);
 
-        // prefix
-        // FIXME?: originally this was for uint64_t, should it be sizeof (size) instead?
-        // Also is usage of uint64_t here really warranted (e.g. dictated by protocol)?
-        zchunk_extend(chunk, (const void *) &size, sizeof (uint64_t));
-        // data
-        zchunk_extend(chunk, (const void *) zframe_data(frame), size);
+        free (alert_desc);
+        cursor = zconfig_next (cursor);
+    }
 
-        zframe_destroy(&frame);
+    zconfig_destroy (&state);
+    zstr_free (&state_file);
+    return 0;
+}
 
-        cursor = (fty_proto_t *) zlistx_next(alerts);
+int
+alert_load_state (zlistx_t *alerts, const char *path, const char *filename)
+{
+    int rv = s_alert_load_state_new (alerts, path, filename);
+    if (rv == -1)
+        rv = s_alert_load_state_legacy (alerts, path, filename);
+
+    return rv;
+}
+
+// right-pad the string with spaces to specified size
+uint8_t *
+s_pad (const char *src, size_t padded_size)
+{
+    uint8_t *padded = (uint8_t *) zmalloc ((padded_size) * sizeof (uint8_t));
+    memcpy (padded, src, strlen (src));
+    for (size_t i = strlen (src); i < padded_size; i++)
+        padded[i] = 0x20; // pad with SPACEs
+
+    return padded;
+}
+
+// save alert state to disk
+// 0 - success, -1 - error
+int
+alert_save_state(zlistx_t *alerts, const char *path, const char *filename, bool verbose)
+{
+    if (!alerts || !path || !filename) {
+        log_error ("cannot save state");
+        return -1;
+    }
+
+    zconfig_t *state = zconfig_new ("root", NULL);
+    fty_proto_t *cursor = (fty_proto_t *) zlistx_first (alerts);
+
+    while (cursor) {
+        fty_proto_print (cursor);
+
+        const char *alert_desc = fty_proto_description (cursor);
+        size_t alert_desc_size = strlen (alert_desc);
+        // new size is the next bigger or equal multiple of 4
+        size_t padded_size = (alert_desc_size + 3) & 0xFFFFFFFC;
```

Comment:
Does this rely on `size_t` being a 32-bit (u)int? Would `(value - (value % 4) + 4)` be more portable?

Reply:
On a second thought, hopefully we won't have strings long enough that this nit would matter.